api/server.js

```diff
@@ -527,69 +530,90 @@ app.postJson('/api/v1/users/:username', function(req, res) {
-          return serverUtils.error({
-            message: 'You do not have permissions to modify this person',
-            code: 401
-          }, req, res);
-        }
-
+    }));
+  const isUpdatingSelf = () =>
+    new Promise((resolve, reject) => auth.getUserCtx(req, (err, userCtx) => {
+      if (err) {
+        reject(err);
+      } else if (userCtx.name === username) {
+        resolve(true);
+      } else {
+        resolve(false);
+      }
+    }));
+  const basicAuthValid = () =>
+    new Promise((resolve, reject) => {
+      const credentials = auth.basicAuthCredentials(req);
+
+      if (!credentials) {
+        resolve(null); // Not attempting basic auth
+      } else if (username !== credentials.username) {
+        // Make sure the Basic Auth credentials are actually for the same person
+        reject({
+          message: 'Basic Auth user does not match the passed username',
+          code: 403
+        });
+      } else {
         auth.validateBasicAuth(credentials, err => {
           if (err) {
             console.error(`Invalid authorization attempt on /api/v1/users/${username}`);
             console.error(err);
-            return serverUtils.error({
-              message: 'Invalid Basic Auth password',
-              code: 401
-            }, req, res);
+            resolve(false); // Incorrect basic auth
+          } else {
+            resolve(true); // Correct basic auth
           }
-
-          // We have valid basic auth and can update ourselves
-          return updateUser(false);
         });
-      } else {
-        // Not trying to update password
-        return updateUser(false);
       }
     });
-  });
+  const isChangingPassword = () => Object.keys(req.body).includes('password');
+
+  Promise.all([hasFullPermission(), isUpdatingSelf(), basicAuthValid(), isChangingPassword()])
+    .then(([fullPermission, updatingSelf, basic, changingPassword]) => {
+
+      if (basic === false) {
+        // If you're passing basic auth we're going to validate it, even if we
+        // technicaly don't need to (because you already have a valid cookie and
+        // full permission).
+        // This is to maintain consistency in the personal change password UI:
+        // we want to validate the password you pass regardless of your permissions
+        throw {
+          message: 'Bad username / password',
+          code: 401
+        };
+      }
+
+      if (fullPermission) {
+        return updateUser(true);
+      }
+
+      if (!updatingSelf) {
+        throw {
+          message: 'You do not have permissions to modify this person',
+          code: 403
+        };
+      }
+
+      if (basic === null && changingPassword) {
+        throw {
+          message: 'You must authenticate with Basic Auth to modify your password',
+          code: 403
+        };
+      }
+
+      return updateUser(false);
+    })
+    .catch(err => {
+      return serverUtils.error(err, req, res);
+    });
```

Comment:
This looks fine. There are several more refactorings you could do if you felt like it, eg: start pushing promises down to the libraries, move more logic out of server.js. But this is a good improvement on its own.